HerPublicWebsite.UnitTests/BusinessLogic/Services/PendingReferralNotificationServiceTests.cs

```diff
@@ -8,23 +15,35 @@ namespace Tests.BusinessLogic.Services;
 [TestFixture]
 public class PendingReferralNotificationServiceTests
 {
+    private Mock<IDataAccessProvider> mockDataAccessProvider;
+    private Mock<CsvFileCreator> mockCsvFileCreator;
     private Mock<IEmailSender> mockEmailSender;
     private PendingReferralNotificationService pendingReferralNotificationService;
     
     [SetUp]
     public void Setup()
     {
+        mockDataAccessProvider = new Mock<IDataAccessProvider>();
+        mockCsvFileCreator = new Mock<CsvFileCreator>();
         mockEmailSender = new Mock<IEmailSender>();
-        pendingReferralNotificationService = new PendingReferralNotificationService(mockEmailSender.Object);
+        pendingReferralNotificationService = new PendingReferralNotificationService(
+            mockDataAccessProvider.Object, 
+            mockCsvFileCreator.Object, 
+            mockEmailSender.Object);
     }
 
     [Test]
-    public void SendPendingReferralNotifications_WhenCalled_CallsSendPendingReferralReportEmail()
+    public async Task SendPendingReferralNotifications_WhenCalled_CallsSendPendingReferralReportEmail()
     {
+        // Arrange
+        mockDataAccessProvider
+            .Setup(dap => dap.GetReferralRequestsBetweenDates(It.IsAny<DateTime>(), It.IsAny<DateTime>()))
+            .ReturnsAsync(new List<ReferralRequest>());
+        
         // Act
-        pendingReferralNotificationService.SendPendingReferralNotifications();
+        await pendingReferralNotificationService.SendPendingReferralNotifications();
         
         // Assert
-        mockEmailSender.Verify(es => es.SendPendingReferralReportEmail());
+        mockEmailSender.Verify(es => es.SendPendingReferralReportEmail(It.IsAny<MemoryStream>()));
```

Comment:
❗❗ You haven't tested the most important part of the code, which is that the correct referrals are included in the CSV.

This is the key part of the logic, so it needs to be covered by unit tests.

You'll have to restructure the code a bit to achieve this. I think the steps you need are:

* Extract the selection of the correct referrals to a `GetReferralRequestsForMonthlyPendingCsv()` on `DataAccessProvider`, which you can then test thoroughly for the correct referrals to include (and not include)
  * To do this, that method will need to know how to find the start of the previous month
  * You should extract this in to a `IDateHelper`, which you can then mock in your tests
  * The implementation of this `DateHelper` should itself be tested
* Mock that new method and the CSV generator in this test class, so that you can confirm that the memory stream being passed in comes from those

Let's have a call to discuss.